### DISCUSSION TITLE
A difficulty to use maps

### DISCUSSION THREAD

**Discussion by claudiosa** [vlang/v#13783] [General]
Hi everyone

I am a newbie in V, so it is my third code in V, and already taking some difficulties with maps. Since to pass it as an argument either to return it.  I want to obtain a map from :

 print('\n MY map: ${creating_a_map(10)}' ) (line 28), where the code is in: 
 https://github.com/claudiosa/CCS/blob/master/v_programming_language/maps_study.v
 but a silly and strange error came:
 $ v run maps_study.v 
maps_study.v:27:4: error: function declarations in script mode should be before all script statements
......
it does not make sense for me yet

In addition, whose types are supported as the keys in V for maps? Only strings? Byte? Runes?

Sorry, I did not scrap for similar questions in this discussion ....
 
Thanks in advance for any help.

Claudio

**Comment by JalonSolov** [comment]
Try moving line 12 up to the end of line 11.

**Reply by claudiosa** [reply]
Jolon

Thanks for your help, but your comment does not make sense. The problem is the type for maps, either as a parameter and as a return.
The layout of types seems a little strange look
$ v run maps_study.v 
maps_study.v:24:9: error: invalid expression: unexpected token `:=`
   22 | fn creating_a_map (size int) [  ] string {}   i16  {
   23 |     
   24 |     my_map := map[  ] string {} i16
      |            ~~
   25 |     mut temp := 'b'
   26 |     mut key := 'a
   
   
   
   I pushed my new code in https://github.com/claudiosa/CCS/blob/master/v_programming_language/maps_study.v
   
   Any additional comment are wellcome, once the maps are not clear yet
   
   Thanks once
   
   
   Claudio

**Comment by JalonSolov** [comment]
Notice that the error message changed.  That was the point of my previous comment.

The main problem with what you have is that the syntax is simply wrong.  You want something more like

`fn creating_a_map(size int) map[string]i16 {`

to return a map with string keys and i16 values.

As for what types you can use as keys... that's actually in the docs, here https://github.com/vlang/v/blob/master/doc/docs.md#maps
```
Maps can have keys of type string, rune, integer, float or voidptr.
```
Enums may also work... not sure.

**Reply by claudiosa** [reply]
Jalon (and community)

Thanks for your tips from  https://github.com/vlang/v/blob/master/doc/docs.md#maps, really I did not see it yet 
My final and as expected code is in:
https://github.com/claudiosa/CCS/blob/master/v_programming_language/maps_study.v
 the details on maps were clarified, the problem is on typing as argument and return. I did some examples to help someone
 
 Thank you ...
 
 
 Claudio
 
 PS: really, V is a very cool language. Modern, fast with syntax clean and elegant!!!!

**Comment by claudiosa** [comment]
Here the final code:

```
import rand

/*
Exploring maps and functions

by CCS
$ v  run maps_study.v 
// Creating a executable
$ v  maps_study.v  -o an_executable.EXE
$ ./an_executable.EXE 
*/

//
fn my_hello () { println("\n Hello World!!!") }

// an array in the argument
fn my_ugly_sum (an_array[] int) int {
	mut n := an_array.len
	mut s := 0
	for i in 0 .. n {
		s = s + an_array[i]
	}
	return s
}

//creating_a_map
fn creating_a_map (size i16)  map [string] i32 {
	mut my_map := map [string] i32 {} 
	mut the_a := "a"
	mut the_b := "b"
	mut key := "a"

	for _ in 0 .. size {
		my_map[key] = rand.i16() % 128  // a random from -128 ... 127
		key = key + the_b + the_a // concat string
	}
    // print('\n My map in the function: ${my_map}' )
	return my_map
 }

// receving and printing a map
fn printing_a_map ( my_map map [string] i32) {
	mut array_of_keys := my_map.keys()
	// println(m.keys()) // ['one', 'two']
	println("\n Printing a map:")
	for i in array_of_keys {
	  print('\n the key[${i}] has the contents:  ${my_map[i]} ' )
	}
    //print('\n My map in the function: ${array_of_keys}' )
 }

fn main() {	
    // mut my_map := map [string] int{} OK
	// mut my_map := map [string] i16{} //OK
	print('\n Creating map from a function: ${creating_a_map(10)}' )
	printing_a_map(creating_a_map(4))
    my_hello() 
	print('\n The ugly sum: ${ my_ugly_sum([4,5,6,7,8,9]) }'  )
	mut my_map_2 := map [string] i16{} //OK
	my_map_2['a'] = 1
	my_map_2['b'] = 2
	print('\n MY map: ${my_map_2}')
	println('\n BYE')
	
}

/************************************************************************/
```

**Reply by Wajinn** [reply]
Good read and to see final code.

**Reply by claudiosa** [reply]
Hi @Wajinn 

Really I had some progress with V. It is a cool language. Please, see some codes with map in 

https://github.com/claudiosa/CCS/blob/master/v_programming_language/maps01_study.v
https://github.com/claudiosa/CCS/blob/master/v_programming_language/maps02_study.v
that was base for this BFS very simple: 
https://github.com/claudiosa/CCS/blob/master/v_programming_language/another_BFS.v

Please, feel free to criticize them, mainly my BFS

Cheers

CCS
